chain/chain/src/store/latest_witnesses.rs

```diff
@@ -0,0 +1,149 @@
+//! for analysis and debugging.
+//! The number of stored witnesses is limited. When the limit is reached
+//! the witness with the oldest height is removed from the database.
+//! At the moment this module is used only for debugging purposes.
+
+use near_primitives::stateless_validation::ChunkStateWitness;
+use near_store::DBCol;
+
+use crate::ChainStoreAccess;
+
+use super::ChainStore;
+
+use borsh::{BorshDeserialize, BorshSerialize};
+use bytesize::ByteSize;
+use near_store::db::LATEST_WITNESSES_INFO;
+use rand::rngs::OsRng;
+use rand::RngCore;
+
+/// Maximum size of the latest witnesses stored in the database.
+const LATEST_WITNESSES_MAX_SIZE: ByteSize = ByteSize::gb(4);
+
+/// Maximum size of a single latest witness stored in the database.
+const SINGLE_LATEST_WITNESS_MAX_SIZE: ByteSize = ByteSize::mb(128);
+
+/// Maximum number of latest witnesses stored in the database.
+const LATEST_WITNESSES_MAX_COUNT: u64 = 60 * 30;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct LatestWitnessesKey {
+    pub height: u64,
+    pub shard_id: u64,
+    /// Each witness has a random UUID to ensure that the key is unique.
+    /// It allows to store multiple witnesses with the same height and shard_id.
+    pub random_uuid: [u8; 16],
+}
+
+impl LatestWitnessesKey {
+    /// `LatestWitnessesKey` has custom serialization to ensure that the binary representation
+    /// starts with big-endan height and shard_id.
+    /// This allows to query using a key prefix to find all witnesses for a given height (and shard_id).
+    fn serialized(&self) -> [u8; 32] {
+        let mut result = [0u8; 32];
+        result[..8].copy_from_slice(&self.height.to_be_bytes());
+        result[8..16].copy_from_slice(&self.shard_id.to_be_bytes());
+        result[16..].copy_from_slice(&self.random_uuid);
+        result
+    }
+}
+
+/// Keeps aggregate information about all witnesses stored in `DBCol::LatestChunkStateWitnesses`.
+/// Used for enforcing limits on the number of witnesses stored in the database.
+#[derive(Debug, Clone, Copy, BorshSerialize, BorshDeserialize, PartialEq, Eq, Default)]
+struct LatestWitnessesInfo {
+    pub count: u64,
+    pub total_size: u64,
+}
+
+impl LatestWitnessesInfo {
+    pub fn is_within_limits(&self) -> bool {
+        self.count <= LATEST_WITNESSES_MAX_COUNT
+            && self.total_size <= LATEST_WITNESSES_MAX_SIZE.as_u64()
+    }
+}
+
+impl ChainStore {
+    /// Saves an observed `ChunkStateWitness` to the database for later analysis and debugging.
+    /// The witness is stored in `DBCol::LatestChunkStateWitnesses`.
+    /// This function does a read-before-write. Don't call it in parallel on the same database,
+    /// or there will be race conditions.
+    pub fn save_lateset_chunk_state_witness(
+        &mut self,
+        witness: &ChunkStateWitness,
+    ) -> Result<(), std::io::Error> {
+        let serialized_witness = borsh::to_vec(witness)?;
+        let serialized_witness_size: u64 =
```

Comment:
I have to serialize the size, so I wanted it to have a clearly defined representation. The exact type of `usize` depends on the architecture, and I didn't want to do that in serialized data.
I guess it doesn't really matter, because no one is going to move the saved witnsses data to a different arch, but it feels more proper to use `u64`.